Queues and Stacks In Python3

TrainedPro opened the pull request:
<!--
Hello! Thank you for contributing to Code Thesaurus!

This template can help you fill out a pull request with all the information 
needed to review it. The text between these exclamation tags are comments 
and won't show up on the pull request. The text after ## are headers. Please
leave the headers unless otherwise noted and follow the instructions to 
fill out each section with as much information as you can to assist the 
reviewer!
-->


## What GitHub issue does this PR apply to?

 <!--
   If you're working on an existing issue, replace "xxxx" below with the issue 
   number. You can change it to say only "Closes #", "Fixes #", or "Resolves #".
   Don't add the word "issue" to it otherwise it won't link correctly.

   If you got here via the edit buttons on the website and/or your changes 
   don't relate to an issue, feel free to delete "Resolves #" and put "None", 
   or note why you chose to make a PR!
   -->

Resolves #536


## What changed and why?

Added queues and stack file describing how the Python functionality would work for it. This was an issue which i was assigned to resolve. The file seems to be as per the requirements.



## Checklist

   <!-- 
   Each - [ ] below is a checkbox that will show up on the PR.
   Either add an X inside the [X], or submit the PR and click the checkboxes.
   -->

- [X] I claimed any associated issue(s) and they are not someone else's
- [X] I have looked at documentation to ensure I made any revisions correctly
- [X] I tested my changes locally to ensure they work
- [ ] (If editing Django) I have added or edited any appropriate unit tests for my changes


## Any additional comments or things to be aware of while reviewing?

   <!-- Please replace this line with any extra information that's helpful -->

I have also changed some names which were incorrect. An example under circular queues is:
`"name": "Data structure that backs the circular queue",` on line 80 where ti was originally simple queue instead of circular queue. This is did not seem to correct it on the website though, when i ran it on my local machine.

TrainedPro committed:
Added queues_stacks.json For Python3

TrainedPro committed:
Merge branch 'codethesaurus:main' into main

geekygirlsarah committed:
Merge branch 'main' into main

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -0,0 +1,388 @@
+      "simple_queue_data_type": {
+        "name": "Queue data type name",
+        "code": [
+          "list"
```
You use a list, which _can_ work for this, but specifically there's [queue.Queue](https://docs.python.org/3/library/queue.html) which seems to be an official queue in Python with fast enqueues and dequeues. It's in the same library as dequeues and PriorityQueues, which you use below. Since lists are described in the `lists.json` file, this should probably be for the built-in queue objects.

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -0,0 +1,388 @@
+      "priority_queue_create_copy_from_list_statement": {
+        "name": "Create a queue from a list of items",
+        "code": [
+          "from queue import PriorityQueue\nimport itertools\n\nqueue = PriorityQueue()\n\npriorities = [1, 2, 3]\n\nmylist = ['One', 'Two', 'Three']\n\nfor priority, item in list(zip(priorities, mylist)):\n\tqueue.put((priority, item))"
```
This is sort of inventing the functionality for this, which isn't what you should do. This should be "not-implemented" as well.

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -0,0 +1,388 @@
+      "priority_queue_destroy_statement": {
+        "name": "Destroy/delete a queue",
+        "code": [
+          "for item in range(queue.qsize()):\n\tqueue.get()"
```
Shouldn't this just be like `del(queue)` ?

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -0,0 +1,388 @@
+      },
+      "stack_import_statement": {
+        "name": "Import statement to add stacks",
+        "not-implemented": "true"
```
I think stacks you can do with a [LIFOqueue](https://docs.python.org/3/library/queue.html) which is essentially a queue but last-in-first-out (which is how a stack works).

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -0,0 +1,388 @@
+      "iterate_create_pointer": {
+        "name": "Create iterator",
+        "code": [
+          "pointer = 0"
```
If iterators/pointers are "not-implemented", then these should be "not-implemented" too. Otherwise they're integers and not iterators.

geekygirlsarah reviewed (changes requested):
This is a lot better! You're getting the idea of how it works. I put comments for some suggestions, and if you fix those, I think I could merge this in!

TrainedPro committed:
Merge branch 'codethesaurus:main' into main

TrainedPro committed:
Resolved All Issues Stated And Any Similar Ones

TrainedPro commented:
All issues stated have been solved along with some others that were similar. Let me know if you would like any further changes.

Thanks!

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -277,13 +273,13 @@
-          "queue.clear()"
+          "stack.clear()"
         ]
       },
```
Is this supposed to be `stack_clear_all` or `queue.clear()` ?

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -277,13 +273,13 @@
-          "mylist = queue"
+          "mylist = stack"
         ]
       },
```
It looks like you're just copying stack over to mylist, but also this is for a queue. I'm a bit confused by this.

TrainedPro replied:
Yes, it seems that I didn't clarify it.

What i did was that where the category was stack, i used the word stack instead of queue, it doesn't matter as it is just a variable name but i just thought it would be more appropriate.

Some place you can see queue used instead of stack even when the name is stack but that was because it was under a queue category, if this is an issue, just let me know as all i really have to do is replace them.

So it shouldn't matter functionally at all. But still, if it matters in terms of presentation, let me know.

geekygirlsarah commented on web/thesauruses/python/3/queues_stacks.json:
```diff
@@ -162,48 +160,48 @@
         "name": "Destroy/delete a stack",
         "code": [
-          "stack.clear()"
+          "del(queue)"
```
Should this be `del(stack)` ?

TrainedPro replied:
I think this accidentally got replaced when i was replacing the other instances, I'll have this resolved as this doesn't fit in.

Please let me know if you want me to use the "stack" variable name for the stack category or should i use queue with all of them.

geekygirlsarah replied:
It could also be `del(lifoqueue)` since that's what a stack is.

geekygirlsarah reviewed (changes requested):
This is a lot closer! I'm wondering if there's a few copy/paste things that you didn't update as I'm seeing multiple mixed up areas between stacks and queues concepts but using the opposite for the code. There's a couple of other comments I had in there too. Run back over those and make sure they line up and I think this might be good.

geekygirlsarah committed:
Merge branch 'main' into main

TrainedPro committed:
Merge branch 'main' into main

TrainedPro committed:
Resolved Issues

Fixed the naming issues between stack and queues as it was unclear before.

TrainedPro commented:
Hello!

I think i fixed all the issues that were presented. I now understand the naming convention you used and renamed the variables according to that. I choose to use stack instead of LifoQueue so it is more simplier and more straight forward as the LifoQueue is being imported and it may cause a bit of confusion.

geekygirlsarah committed:
Merge branch 'main' into main

geekygirlsarah reviewed (changes requested):
So I'm starting to realize there's some conflicts in this file. Things like:

* dequeue has .append() and .appendLeft() to add to it but Queue/SimpleQueue/LifoQueue has .put()
* dequeue has .pop() and .popleft() but Q/SQ/LQ has .get()
* dequeue has .clear() and .copy() but Q/SQ/LQ doesn't have an equivalent
* dequeue is iterable, Q/SQ/LQ aren't

So parts of this file are correct and parts aren't, and it's entirely dependent on the type of queue you build with. Which probably means there needs to be some comments in there to explain what works with each, or you need more elaborate code blocks. Something like, for example:
```
    "queue_enqueue_an_item": {
        "name": "Enqueue an item (add to end)",
        "code": [
          "deque.append('Four')",
          "queue.put('Four')",
          "lifoqueue.put('Four')",
          "simplequeue.put('Four')"
        ]
      },
```

That way the user can compare all of the objects and see what works. For things like clearing a queue, you might need something like:

```
     "queue_clear_all": {
        "name": "Clear out all queue items",
        "code": [
          "deque.clear()"
        ],
        "comment": "Queues, SimpleQueues, and LifoQueues don't have a clear function."
      },
```

I think it still needs another look over. I appreciate you continuing to work on this and making it better, but it looks like as you've corrected it, a few other confusions have popped up. Do you mind fixing these?

TrainedPro commented:
No worries, I'll go through it and I'll do these changes.

TrainedPro committed:
Merge branch 'codethesaurus:main' into main

TrainedPro commented:
Ok, so i think i implemented all the changes other than the `.copy()` as in Python, that is not too widely used. It is much more common to copy it by assigning it to another variable, which is what i have used.

TrainedPro committed:
Resolved Conflicts

TrainedPro committed:
Resolved Conflicts

TrainedPro committed:
Resolved Conflicts

TrainedPro committed:
Resolved Conflicts

TrainedPro committed:
Resolved Conflicts

geekygirlsarah committed:
Merge branch 'main' into main

geekygirlsarah committed:
Merge branch 'main' into main

geekygirlsarah commented:
> Ok, so i think i implemented all the changes other than the `.copy()` as in Python, that is not too widely used. It is much more common to copy it by assigning it to another variable, which is what i have used.

I think Python shallow copies it. (So like the variable points to the object in memory, and `var2 = var1` would just point them both to the same memory location. It's not a real copy since if you do something to the first one, you've also done it to the second.)

geekygirlsarah reviewed (approved):
I think you resolved the things I noticed, and if there's other problems, someone can file an issue with it.

Thanks for your patience and endurance while working on this! I'll merge it in. Congrats on your first Code Thesaurus contribution! 🎉 

Have a great weekend!
